cli/democluster: TestTransientClusterSimulateLatencies failed (nil panic in maybeRejectUnsafeProposalLocked)

cli/democluster.TestTransientClusterSimulateLatencies [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5549523?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5549523?buildTab=artifacts#/) on master @ [7af6c0d3e2085cafda90d5f0b15d06f6a360f154](https://github.com/cockroachdb/cockroach/commits/7af6c0d3e2085cafda90d5f0b15d06f6a360f154):


```
=== RUN   TestTransientClusterSimulateLatencies
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/b296432f77c11d14af9987c5728df5a9/logTestTransientClusterSimulateLatencies3489558533
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTransientClusterSimulateLatencies
    panic.go:1047: -- test log scope end --
    panic.go:1047: runtime error: invalid memory address or nil pointer dereference
        goroutine 10493 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x92
        panic({0x5734e80, 0x87004f0})
        	GOROOT/src/runtime/panic.go:1047 +0x266
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).maybeRejectUnsafeProposalLocked(0xc00475aab8, {0x64f8bd8, 0xc00526e390}, {0x0, 0x0}, 0xc006450b40)
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_proposal_buf.go:690 +0x41f
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).FlushLockedWithRaftGroup(0xc00475aab8, {0x64f8bd8, 0xc00526e390}, {0x0, 0x0})
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_proposal_buf.go:428 +0x4d5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).FlushLockedWithoutProposing(0xc00475aab8, {0x64f8bd8, 0xc00526e390})
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_proposal_buf.go:921 +0x65
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processRaft.func2.1(0xc00475a400)
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711 +0x97
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*storeReplicaVisitor).Visit(0xc009abd2c0, 0xc01334d6b8)
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store.go:416 +0x25c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).VisitReplicas(0xc001aa3500, 0xc01334d6b8, {0x0, 0x0, 0x0})
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store.go:2633 +0xc5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processRaft.func2()
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:709 +0x85
        github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc005737be0)
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:107 +0x1f
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc000bc26c0, {0x64f8b68, 0xc00013a000})
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:558 +0x3ae
        github.com/cockroachdb/cockroach/pkg/server.(*Server).Stop(0xc001082400)
        	github.com/cockroachdb/cockroach/pkg/server/server.go:1608 +0x4a
        github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc00087f1d0)
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:107 +0x1f
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc000bc2480, {0x64f8b68, 0xc00013a000})
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:558 +0x3ae
        github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).Close(0xc000412ff0, {0x64f8b68, 0xc00013a000})
        	github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:812 +0x4c
        github.com/cockroachdb/cockroach/pkg/cli/democluster.TestTransientClusterSimulateLatencies(0xc001098340)
        	github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster_test.go:256 +0x846
        testing.tRunner(0xc001098340, 0x5f48808)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
--- FAIL: TestTransientClusterSimulateLatencies (11.46s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTransientClusterSimulateLatencies.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-16989

cli/democluster.TestTransientClusterSimulateLatencies [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5570584?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5570584?buildTab=artifacts#/) on master @ [a611410e3babe97a14152381b89060e1bafa2e27](https://github.com/cockroachdb/cockroach/commits/a611410e3babe97a14152381b89060e1bafa2e27):


```
=== RUN   TestTransientClusterSimulateLatencies
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/b296432f77c11d14af9987c5728df5a9/logTestTransientClusterSimulateLatencies2570579511
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTransientClusterSimulateLatencies
    panic.go:1047: -- test log scope end --
    panic.go:1047: runtime error: invalid memory address or nil pointer dereference
        goroutine 82 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x92
        panic({0x573f4a0, 0x8711500})
        	GOROOT/src/runtime/panic.go:1047 +0x266
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).maybeRejectUnsafeProposalLocked(0xc004a572b8, {0x6508718, 0xc0041f98f0}, {0x0, 0x0}, 0xc0052c0000)
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_proposal_buf.go:690 +0x41f
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).FlushLockedWithRaftGroup(0xc004a572b8, {0x6508718, 0xc0041f98f0}, {0x0, 0x0})
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_proposal_buf.go:428 +0x4d5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).FlushLockedWithoutProposing(0xc004a572b8, {0x6508718, 0xc0041f98f0})
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_proposal_buf.go:921 +0x65
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processRaft.func2.1(0xc004a56c00)
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711 +0x97
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*storeReplicaVisitor).Visit(0xc00e003a10, 0xc012cdf6b8)
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store.go:416 +0x25c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).VisitReplicas(0xc0007ec000, 0xc012cdf6b8, {0x0, 0x0, 0x0})
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store.go:2633 +0xc5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processRaft.func2()
        	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:709 +0x85
        github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc003311660)
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:107 +0x1f
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc00038ccf0, {0x65086a8, 0xc000080048})
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:558 +0x3ae
        github.com/cockroachdb/cockroach/pkg/server.(*Server).Stop(0xc000b25b00)
        	github.com/cockroachdb/cockroach/pkg/server/server.go:1608 +0x4a
        github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc0009722b0)
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:107 +0x1f
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc00038cb40, {0x65086a8, 0xc000080048})
        	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:558 +0x3ae
        github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).Close(0xc00026d950, {0x65086a8, 0xc000080048})
        	github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:812 +0x4c
        github.com/cockroachdb/cockroach/pkg/cli/democluster.TestTransientClusterSimulateLatencies(0xc000603a00)
        	github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster_test.go:256 +0x846
        testing.tRunner(0xc000603a00, 0x5f54888)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
--- FAIL: TestTransientClusterSimulateLatencies (3.72s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTransientClusterSimulateLatencies.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_proposal_buf.go
+++ b/pkg/kv/kvserver/replica_proposal_buf.go
@@ -567,6 +567,13 @@ func (b *propBuf) FlushLockedWithRaftGroup(
 func (b *propBuf) maybeRejectUnsafeProposalLocked(
 	ctx context.Context, raftGroup proposerRaft, p *ProposalData,
 ) (rejected bool) {
+	if raftGroup == nil {
+		// If we do not have a raft group, we won't try to propose this proposal.
+		// Instead, we will register the proposal so that it can be reproposed later
+		// with a raft group. Wait until that point to determine whether to reject
+		// the proposal or not.
+		return false
+	}
 	switch {
 	case p.Request.IsSingleRequestLeaseRequest():
 		// Handle an edge case about lease acquisitions: we don't want to forward
@@ -706,9 +713,6 @@ func (b *propBuf) maybeRejectUnsafeProposalLocked(
 // leaderStatusRLocked returns the rangeLeaderInfo for the provided raft group,
 // or an empty rangeLeaderInfo if the raftGroup is nil.
 func (b *propBuf) leaderStatusRLocked(ctx context.Context, raftGroup proposerRaft) rangeLeaderInfo {
-	if raftGroup == nil {
-		return rangeLeaderInfo{}
-	}
 	leaderInfo := b.p.leaderStatusRLocked(raftGroup)
 	// Sanity check.
 	if leaderInfo.leaderKnown && leaderInfo.leader == b.p.getReplicaID() &&

